### PR TITLE
Implement hessian autodiff calculation

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
restrictions:
- cannot jit
- only the `forward_common` interface has its hessian calculation. not for `forward_common_lower`.
- may give nan when nall == nloc. specifically when nloc==1

also fix bug in pt: transform_output. The output shape will be wrong when the dimension of output variable is larger than 1.

### CODE REVIEW

File: deepmd/pt/model/model/make_hessian_model.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                    hess_keys.append(kk)
+            # result dict init by empty lists
+            res = {get_hessian_name(kk): [] for kk in hess_keys}
+            # loop over variable
+            for kk in hess_keys:
+                vdef = fdef[kk]
+                vshape = vdef.shape
+                vsize = 1
+                for ii in vshape:
+                    vsize *= ii
```
Comment (njzjz) [inline deepmd/pt/model/model/make_hessian_model.py:144]: `vsize = math.prod(vshape)` or `vsize = np.prod(vshape)`

File: deepmd/pt/model/model/make_hessian_model.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                # loop over frames
+                for ii in range(nf):
+                    icoord = coord[ii]
+                    iatype = atype[ii]
+                    ibox = box[ii] if box is not None else None
+                    ifparam = fparam[ii] if fparam is not None else None
+                    iaparam = aparam[ii] if aparam is not None else None
+                    # loop over all components
+                    for idx in range(vsize):
```
Comment (njzjz) [inline deepmd/pt/model/model/make_hessian_model.py:153]: Will it be slow when using `for ... in` twice?

Comment (wanghan-iapcm) [reply deepmd/pt/model/model/make_hessian_model.py:153]: I haven't benchmarked, but i guess there should not be a significant difference.

File: deepmd/pt/model/model/make_hessian_model.py (hunk 1 above)
Comment (wanghan-iapcm) [reply deepmd/pt/model/model/make_hessian_model.py]: for the sake of jit... now we use a more elegant way

**Review by njzjz — APPROVED** [review]

**Review by iProzd — APPROVED** [review]

**Pull request merged**